lib/Sema/CSStep.h

```diff
@@ -607,21 +630,34 @@ class TypeVariableStep final : public BindingStep<TypeVarBindingProducer> {
         << InitialBindings.size() << " initial bindings\n";
   }
 
+private:
+  /// Checks whether any of the attempted bindings so far have produced a
+  /// solution without needing to add fixes or bind to an unavailable decl.
+  bool solvedAnyWithoutFixes() const {
+    if (!AnySolved)
+      return false;
+
+    // Check if any of the bindings introduced fixes or bound unavailable
+    // overloads.
+    auto scoreDelta = *BestScore - OriginalScore;
+    return scoreDelta.Data[SK_Fix] == 0 && scoreDelta.Data[SK_Unavailable] == 0;
+  }
+
 protected:
   bool attempt(const TypeVariableBinding &choice) override;
 
   bool shouldSkip(const TypeVariableBinding &choice) const override {
     // If this is a defaultable binding and we have found solutions,
     // don't explore the default binding.
-    return AnySolved && choice.isDefaultable();
+    return solvedAnyWithoutFixes() && choice.isDefaultable();
   }
 
   /// Check whether attempting type variable binding choices should
   /// be stopped, because optimal solution has already been found.
   bool shouldStopAt(const TypeVariableBinding &choice) const override {
     // If we were able to solve this without considering
     // default literals, don't bother looking at default literals.
-    return AnySolved && choice.hasDefaultedProtocol() &&
+    return solvedAnyWithoutFixes() && choice.hasDefaultedProtocol() &&
```

Comment:
Instead of introducing separate scoring, could you try doing this -  https://github.com/apple/swift/pull/30006/files#diff-0a3b5fe8ff3dd168f45ad89d96d2ac3aR623. This is was a part of a PR had had to revert but it might be useful still... Maybe we should always

Reply:
@xedin Tried locally, and yup that works instead of `solvedAnyWithoutFixes()` for all three methods.

My main reason for going with the scoring check was to allow us to explore roughly the same set of bindings as we would have done without being in diagnostic mode, and still skip bindings when we didn't need to add a fix along a particular path.

Should I switch over to using `shouldAttemptFixes` for this and `shouldStopAfter`?

Reply:
My preference would be do use `shouldAttemptFixes` if that works and not optimize number of bindings for now.